### PR TITLE
Add block volume support to internal provisioners.

### PR DESCRIPTION
**Pull request by jsafrane**
**What this PR does / why we need it**:
Internal provisioners now create filesystem PVs when block PVs are requested. This leads to unbindable PVCs.

In this PR, volume plugins that support block volumes provision block PVs when block is requested. All the other provisioners return clear error in `kubectl describe pvc`:

```
Events:
  Type     Reason              Age               From                         Message
  ----     ------              ----              ----                         -------
  Warning  ProvisioningFailed  7s (x2 over 18s)  persistentvolume-controller  Failed to provision volume with StorageClass "standard": kubernetes.io/cinder does not support block volume provisioning

```

```release-note
AWS EBS, Azure Disk, GCE PD and Ceph RBD volume plugins support dynamic provisioning of raw block volumes.
```

cc @kubernetes/vmware for vsphere changes
cc @andyzhangx for Azure changes

/assign @copejon @mtanino

### CODE REVIEW

**Comment by jsafrane** [comment]
e2e tests will follow

**Review by andyzhangx — APPROVED** [review]
/lgtm 
on azuredisk & azurefile part

**Comment by jsafrane** [comment]
Thanks for the quick review, however,  
/lgtm cancel

I need the others reviewer too :-)

**Comment by jsafrane** [comment]
/approve
stupid bot!

**Comment by jsafrane** [comment]
Added e2e test. I ran it manually on GCE and AWS. It will be run in alpha test suite(s), since it depends on alpha feature.

**Comment by childsb** [comment]
This seems like a feature, but its really a bug fix:

1) we added block volume support for SOME of the plugins
2) The plugins which we didn't explicitly get block volume support don't know that they didn't
3) the provisioners for the volume types which dont have block volume support still provision regular volumes when the block volume flag is set in the storage class.

ideally these volume types wouldn't even provision volumes, but enforcing that may be more difficult then enabling them to provision block volumes. 

would like this in as a fix for 1.11

**Comment by msau42** [comment]
GCE and E2E changes LGTM

**Comment by msau42** [comment]
> This seems like a feature, but its really a bug fix
The other major bug fix included here is that for the plugins that actually added block support, they didn't actually create PVs with the correct volumeMode.

File: pkg/volume/vsphere_volume/vsphere_volume.go
Comment (mkimuram) [inline pkg/volume/vsphere_volume/vsphere_volume.go]: I'm afraid that vSphere doesn't seem to support BlockVolume, so shouldn't it return error here?

**Comment by mkimuram** [comment]
In addition, it might be out of the scope of this issue, but how should external provisioners handle this?
(We have no control for external provisioners, so we can't add similar codes to all external provisioners.)

Do we need some interface to check supportability of plugins?
I found that internal provisioner checks SupportsMountOption() before Provision(), so I guess that we would be better to do similar check around here and external provisioner, in the future.
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1421

File: pkg/volume/vsphere_volume/vsphere_volume.go
Comment (jsafrane) [reply pkg/volume/vsphere_volume/vsphere_volume.go]: Fixed, now it returns the same error as other plugins that don't support block.

**Comment by jsafrane** [comment]
> Do we need some interface to check supportability of plugins?

I think that it's enough to have a check in `Provision()` call in each volume plugin. They're responsible to provision something that corresponds to given PVC, incl. block volume support. I don't like adding extra call to check support of every tiny PVC detail.

MountOptions are quite specific that they're just copied from StorageClass to PV during provisioning, so the volume plugin does not need to care in `Provision()`, they're used during mount in kubelet.

> In addition, it might be out of the scope of this issue, but how should external provisioners handle this?

Traditionally, we leave this to users / cluster admins. It's their responsibility to run their clusters with compatible components, i.e. with the provisioners that support block. And that's what are alpha/beta feature gates good for. I know it's not ideal, however, there is no direct call-like interface between Kubernetes and external provisioners.

**Comment by screeley44** [comment]
aws and e2e LGTM

**Comment by mkimuram** [comment]
To discuss similar check for external provisioner, I made PR for it.
- [WIP] Add block volume support check to external provisioners
  https://github.com/kubernetes-incubator/external-storage/pull/787

As described there, similar issue to CSI also happens in flex volume plugin.

**Comment by jsafrane** [comment]
I got reviews from all volume plugins (@rootfs for Cinder and RBD)

/sig storage

**Comment by jsafrane** [comment]
> Action required: This pull request must have the status/approved-for-milestone label applied by a SIG maintainer.

@childsb @saad-ali, can you please make sure this is merged in 1.11?

**Pull request merged**
